ddp-workspace/projects/ddp-sdk/src/lib/services/activity/activitySuggestionBuilder.service.ts

```diff
@@ -107,61 +121,47 @@ export class ActivitySuggestionBuilder {
             }
         };
 
-        let suggestionsLeft = suggestions.filter(hasAnyMatch);
+        const suggestionsMatched = suggestions.filter(hasAnyMatch);
+        const resultForStartWith = this.findMatches(suggestionsMatched, indexOfStartWithMatch);
+        const resultForStartOfWord = this.findMatches(resultForStartWith.leftover, indexOfStartOfWordMatch);
+        const resultForWithinWord = this.findMatches(resultForStartOfWord.leftover, indexOfFirstMatch);
+        const startsWithSuggestions = resultForStartWith.matches;
+        const startOfWordSuggestions = resultForStartOfWord.matches;
+        const suggestionsWithinWords = resultForWithinWord.matches;
 
-        const startsWithSuggestions = suggestionsLeft.reduce((accumulator: Array<TextSuggestion>, suggestion: string) => {
-            if (hasStartWithMatch(suggestion)) {
-                accumulator.push({
-                    value: suggestion,
-                    matches: [{
-                        offset: 0,
-                        length
-                    }]
-                });
-            }
-            return accumulator;
-        }, []);
-
-        suggestionsLeft = suggestionsLeft.filter((suggestion: string) => !hasStartWithMatch(suggestion));
-
-        const startOfWordSuggestions = suggestionsLeft.reduce((accumulator: Array<TextSuggestion>, suggestion: string) => {
-            const offset = hasStartOfWordMatch(suggestion);
-            if (offset !== -1) {
-                accumulator.push({
-                    value: suggestion,
-                    matches: [{
-                        offset,
-                        length
-                    }]
-                });
-            }
-            return accumulator;
-        }, []);
+        startsWithSuggestions.sort(compareOffsetAndText);
+        startOfWordSuggestions.sort(compareOffsetAndText);
+        suggestionsWithinWords.sort(compareOffsetAndText);
 
-        suggestionsLeft = suggestionsLeft.filter((suggestion: string) => hasStartOfWordMatch(suggestion) === -1);
+        return [
+            ...startsWithSuggestions,
+            ...startOfWordSuggestions,
+            ...suggestionsWithinWords
+        ];
+    }
 
-        const suggestionsWithinWords = suggestionsLeft.reduce((accumulator: Array<TextSuggestion>, suggestion: string) => {
-            const offset = indexOfFirstMatch(suggestion);
-            if (offset !== -1) {
+    private findMatches(suggestions: Array<string>, matcher: (string) => number): MatchResult {
+        const leftover = [];
+        const matches = suggestions.reduce((accumulator: Array<TextSuggestion>, suggestion: string) => {
+            const offset = matcher(suggestion);
+            if (offset >= 0) {
                 accumulator.push({
                     value: suggestion,
                     matches: [{
                         offset,
                         length
```

Comment:
I've fixed the test and but if you run `ng test ddp-sdk` you can observe that some of them are still failing. This happens due to this variable. I've debugged the code and found out that we have `length` variable in the global scope and it equals 0. To fix this and build the correct SuggestionMatch result, we need to pass `length` explicitly as a third parameter in the `findMatches` in the `buildTextSuggestions` call.

Reply:
Oops! I missed it while refactoring. Thanks for the catch!

Reply:
Fixed! I just moved it back to a closure so it can access the `length` variable.